main.tf

```
@@ -73,7 +79,12 @@ locals {
     attributes  = local.id_context.attributes
   }
 
-  generated_tags = { for l in keys(local.tags_context) : title(l) => local.tags_context[l] if length(local.tags_context[l]) > 0 }
+  generated_tags = {
+    for l in keys(local.tags_context) :
+    local.generated_tag_name_case == "title" ? title(l) : (
+      local.generated_tag_name_case == "lower" ? lower(l) : upper(l)
+    ) => local.tags_context[l] if length(local.tags_context[l]) > 0
+  }
```

@osterman I think we should throw an error on unexpected values, or at least default to the default. What do you think?

```suggestion
  locals {
    valid_tag_name_cases = {
      lower = "lower"
      title = "title
      upper = "upper"
    }
  }
  
  generated_tags = {
    for l in keys(local.tags_context) :
    // lookup in valid_tag_name_cases will throw error if there is no match to a key
    valid_tag_name_cases[local.generated_tag_name_case] == "upper" ? upper(l) : (
      local.generated_tag_name_case == "lower" ? lower(l) : title(l)
    ) => local.tags_context[l] if length(local.tags_context[l]) > 0
  }
```

---

There's no official way to throw exceptions, and I'm not crazy about the hacks which call exec on something like "this is not expected" (assuming that there's no binary called "this is not expected").

---

```hcl
  validation {
    condition     = contains(["lower", "title", "upper"], var.generated_tag_name_case)
    error_message = "Allowed values only `lower`..."
  }
 ```
 @Nuru @osterman what do you think?

---

Ok, that makes sense!

---

@osterman @SweetOps I would be happy with validations except they are not supported in TF 0.12 and I do not think they are worth dropping TF 0.12 support for.

Because of that and since there is no official way to raise an error, I proposed the map lookup. It will fail with a possibly helpful error message. Something like
```
    | local.valid_tag_name_cases is object with 3 attributes
    | local.generated_tag_name_case is "base"
The given key does not identify an element in this collection value.
```
I think that is not great, but not terrible either. At least we will not let a bad value through, and we can maintain TF 0.12 support.

---

But we have no need to maintain 0.12 support in *new* versions. Anyone on 0.12 can surely continue to use an older version of 0.12 without the case transformations. 

0.12 will soon be > 3 versions behind (0.15 will drop any day now).  

I argue we should be using validations in all of our modules and use first-class support in the language.

---

@osterman OK to forcing upgrade to TF 0.13 and using first-class validations. Just in case that all fails, though, I would still prefer the uncaught error result in the published default behavior rather than something else:
```suggestion
  generated_tags = {
    for l in keys(local.tags_context) :
    local.generated_tag_name_case == "upper" ? upper(l) : (
      local.generated_tag_name_case == "lower" ? lower(l) : title(l)
    ) => local.tags_context[l] if length(local.tags_context[l]) > 0
  }
```